### DISCUSSION TITLE
Best way to show a nested object inside a datagrid

### DISCUSSION THREAD

**Discussion by prakhargupta1** [mui/toolpad#1419] [Help and support] [answered]
I want to do [this](https://codesandbox.io/s/romantic-dream-4qi63b?file=/src/App.js) but in Toolpad. It has nested object and is easier to add in the table the pro-code way. 

In Toolpad, I am expected to first transform the nested object to a basic object. 
Retool also does it this way: https://community.retool.com/t/nested-json-to-table/8221

My question is: As nested objects are quite common in big data sets. Is transformation the best way to handle it? Or, like pro-code can we have table cell abstractions inside the Table row as well? It will remove the need to write unnecessary transformation code.

**Comment by bharatkashyap** [comment]
> I want to do [this](https://codesandbox.io/s/romantic-dream-4qi63b?file=/src/App.js)

I can't seem to understand what the requirement is since it only seems to show this:
<img width="1190" alt="Screenshot 2022-12-06 at 12 02 04 PM" src="https://user-images.githubusercontent.com/19550456/205838312-6ac31b3d-5791-493e-872c-0b71a6f78d22.png">

**Reply by prakhargupta1** [reply]
Please retry, missed saving it.

**Reply by bharatkashyap** [reply, marked as answer]
I would think that a transform on the query is the best way to accomplish this.

> It will remove the need to write unnecessary transformation code

I think we could at some point even auto-suggest the transformation code instead of writing a lot more code to build UI to accomplish tasks that could be done via a single-line of transformation code.

**Comment by cherniavskii** [comment]
This might be too late, but I think it's still worth answering the question :)
Yes, data transformation is probably the simplest option.

There's one more way of dealing with nested data - you can use [`valueGetter`](https://mui.com/x/react-data-grid/column-definition/#value-getter):
```tsx
const columns = [
  {
    field: "companyName",
    valueGetter: ({ row }) => row.data.company.name,
  }
]
```
If you want to also make the column editable - you will need to provide [`valueSetter`](https://mui.com/x/react-data-grid/editing/#value-parser-and-value-setter):
```tsx
const columns = [
  {
    field: "companyName",
    editable: true,
    valueGetter: ({ row }) => row.data.company.name,
    valueSetter: ({ row, value }) => {
  	  return {
  	    ...row,
  	    data: {
  	      ...row.data,
  	      company: {
  	        ...row.data.company,
  	        name: value,
  	      }
  	    }
	  }
    }
  }
]
```

Here's a demo using `DataGrid` with the data you provided: https://codesandbox.io/s/mui-datagrid-forked-7bgp5e?file=/src/App.js
In this demo, I have abstracted the `valueGetter`/`valueSetter` repetitive code to a `nested` helper function and used `lodash` to simplify things.

Hope this answers the question, let me know if there are any questions left :)

**Reply by Janpot** [reply]
Thanks, the main question here will be about how to define it in a [low-code](https://github.com/mui/mui-toolpad/issues/1776) way. The solution will naturally use `valueGetter` under the hood.
